### PR TITLE
Initial port of auto-find-k

### PR DESCRIPTION
**Pull request by cjnolet**
This is a port of https://github.com/rapidsai/cuml/issues/818 (originally from NVGraph) which uses the Calinski-Harabasz score 
to find the optimal value of k. 

Todo:
- [x] create histogram of cluster sizes
- [x] add googletests 
- [x] expose public API

Closes https://github.com/rapidsai/raft/issues/825

cc @jeaton32

### CODE REVIEW

**Comment by cjnolet** [comment]
rerun tests

File: cpp/include/raft/cluster/detail/kmeans_auto_find_k.cuh [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+                                                                labels.view(),
+                                                                residual,
+                                                                n_iter);
+
+      detail::countLabels(
+        handle, labels.data_handle(), clusterSizes.data_handle(), n, mid, workspace);
+
+      resultsView[mid]           = residual[0];
+      clusterDispertionView[mid] = raft::stats::cluster_dispersion(
+        handle, centroids_const_view, cluster_sizes_view, std::nullopt, n);
```
Comment (benfred) [inline cpp/include/raft/cluster/detail/kmeans_auto_find_k.cuh:168]: This block of code appears 3x in this function  (here, and for calculating left/right values above)- maybe we should refactor to remove the duplicate code? I think we could pull out to a lambda fairly easily, and then call that

like 
```c++
auto calculate_dispersion = [&](int idx) { 
... 
};
calculate_dispersion(mid);
```

**Review by benfred — COMMENTED** [review]
This looks really good! 

Some minor comments:

File: cpp/include/raft/cluster/detail/kmeans_auto_find_k.cuh (hunk 1 above)
Comment (cjnolet) [reply cpp/include/raft/cluster/detail/kmeans_auto_find_k.cuh:168]: Good point! I'll fix this.

**Review by benfred — APPROVED** [review]
lgtm!

**Pull request merged**